api/pubsub.py

```diff
@@ -82,8 +86,37 @@ async def unsubscribe(self, sub_id):
             if sub is None:
                 raise ValueError(f"Invalid subscription id: {sub_id}")
             self._subscriptions.pop(sub_id)
+            self._filters.pop(sub_id)
             await sub.unsubscribe()
 
+    def pubsub_event_filter(self, sub_id, msg):
```

Comment:
One issue is that the logic should not have knowledge of the schema, or it will need to be updated all the time to be kept in sync and that's not maintainable. Instead, the data used for filtering should be provided alongside the filters and the logic could go through it recursively.